app/coffee/AuthHandler.coffee

```diff
@@ -0,0 +1,64 @@
+bcrypt = require("bcrypt")
+crypto = require("crypto")
+logger = require("logger-sharelatex")
+
+{ db } = require("./mongojs")
+Errors = require("./Errors")
+
+TOKEN_TTL_MS = 60 * 60 * 1000
+
+module.exports = AuthHandler =
+	login: (email, password, callback) ->
+		db.users.findOne {email: email.toLowerCase()}, (err, user) ->
+			if err
+				return callback(err)
+			if !user
+				return callback(new Errors.AuthenticationError("user not found"))
+			if !user.hashedPassword
+				return callback(new Errors.AuthenticationError("user does not have a password"))
+			bcrypt.compare password, user.hashedPassword, (err, match) ->
+				if err
+					return callback(err)
+				if !match
+					return callback(new Errors.AuthenticationError("incorrect password"))
+				callback(null, user._id.toString())
+
+	oneTimeLogin: (email, token, callback) ->
+		now = new Date()
+		normalizedEmail = email.toLowerCase()
+		query = {
+			email: normalizedEmail
+			token: token
+			expiresAt: { $gt: now }
+			usedAt: { $exists: false }
+		}
+		update = {
+			$set: { usedAt: now }
+		}
+		db.tokens.findAndModify { query: query, update: update }, (err, token) ->
+			if err
+				return callback(err)
+			if !token
+				return callback(new Errors.AuthenticationError("invalid token"))
+			db.users.findOne { email: normalizedEmail }, (err, user) ->
+				if err
+					return callback(err)
+				if !user
+					return callback(new Errors.AuthenticationError("user not found"))
+				callback(null, user._id.toString())
+
+	generateOneTimeLoginToken: (email, callback) ->
+		logger.info({ email }, "Generating one-time login token")
+		now = new Date()
+		expiresAt = new Date(now.getTime() + TOKEN_TTL_MS)
+		token = crypto.randomBytes(32).toString('hex')
```

Comment:
I think bcrypt may be over the top here, because these are long and random tokens (as opposed to user passwords that may be short and reused on other sites) protected by a rate limiter, and they are short lived (these tokens should only last an hour or two).

I think timing attacks are a concern, however, because we are looking up the whole token in the database (together with the email address). An approach we have taken elsewhere is to use a database-generated ID and a random token together. The ID part guarantees uniqueness, and you can use it for the lookup; then you can then do a timing-safe compare on the token (see for example how we do it [in web](https://github.com/overleaf/web/blob/1c30916030261d8bb295e28b10cd632033b80cb3/app/src/Features/TokenAccess/TokenAccessHandler.js#L98-L101)).

(If we only allowed one token per email address, we could look up on the email only and then do a timing-safe compare on the token, but as mentioned above, I think limiting it to one token per email address may not be a very good UX.)

Reply:
Thanks for the timing considerations. I hadn't thought of that. Would it work if I combine your two approaches?

* Split the token in two
* Use the first part of the token in the db query
* Compare the whole token using `crypto.timingSafeEqual()`

If I understand correctly, this would limit any timing attack to discovering only the first half of the token.

Reply:
I think that would work, yes. If the tokens already have an ID from Mongo, I think using it would make sense... but it would require some changes downstream to pass both parameters. In the presence of a fairly low rate limiter, I think it's probably OK as-is.